### PR TITLE
Add Harvester provider

### PR DESCRIPTION
**Pull request by futuretea**
Signed-off-by: futuretea <Hang.Yu@suse.com>

1. add harvester provider
2.  add harvester filter to k3s and rke2  (make sure detect harvester correctly if harvester is based on k3s/rke2 )

**Related issues**
harvester/harvester#861

### CODE REVIEW

File: providers/aks.go [hunk 1]
```diff
@@ -9,7 +9,15 @@ import (
 
 const AKS = "aks"
 
-func IsAKS(ctx context.Context, k8sClient kubernetes.Interface) (bool, error) {
+var _ = &AKSProvider{}
```
Comment (guangbochen) [inline providers/aks.go:12]: why do we define the empty var here?

Comment (futuretea) [reply providers/aks.go]: updated

**Review by guangbochen — APPROVED** [review]
LGTM

File: providers/provider.go [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package providers
+
+import (
+	"context"
+
+	"k8s.io/client-go/kubernetes"
+)
+
+// IsProvider is the interface all providers need to implement
+type IsProvider func(ctx context.Context, k8sClient kubernetes.Interface) (bool, error)
```
Comment (gitlawr) [inline providers/provider.go:10]: With the interface changes this is not used anymore

Comment (futuretea) [reply providers/provider.go]: fixed

**Review by gitlawr — APPROVED** [review]
LGTM

**Review by dramich — CHANGES_REQUESTED** [review]
The intention of the map is so we don't end up in an order matters situation. Currently, the rke check will pass if it detects a windows node annotation. RKE2 and k3s can do a similar check so we fall through to harvester as being the provider if they end up running first.

**Comment by guangbochen** [comment]
is it possible for us to return both harvester and rke2, or maybe we should call it `harvester:rke2` or `harvester-rke2` if we support manage both VM and container cluster view. cc @yasker

**Comment by dramich** [comment]
> is it possible for us to return both harvester and rke2, or maybe we should call it `harvester:rke2` or `harvester-rke2` if we support managing both VM and container cluster view. cc @yasker

The rke windows provider uses https://github.com/rancher/kubernetes-provider-detector/blob/master/providers/rke-windows.go#L10 `"rke.windows"` so we should follow that pattern. But it's up to your team how detailed you need the provider to be. If harvester handles rke2 vs k3s differently then include that info.

**Comment by yasker** [comment]
@dramich  Does `rke2` or `k3s` have an impact on how integration as Kubernetes cluster works? I am not sure if we should expose that information.

**Comment by dramich** [comment]
> @dramich Does `rke2` or `k3s` have an impact on how integration as Kubernetes cluster works? I am not sure if we should expose that information.

Yes, but if your provider ends up being `harvester.rke2` then it won't matter to rancher as the logic that checks for rke2 specifically won't apply. If the backing cluster isn't important to harvester, meaning you never have branching logic based on rke2 vs k3s, then don't expose it, but I don't think having that data now would be bad as it might be needed later.

**Comment by dramich** [comment]
@yasker Thinking about it I think you should add the underlying cluster so the UI can do the right thing when enabling monitoring on the cluster. k3s vs rke2 has different monitoring chart options. cc @aiyengar2

**Comment by yasker** [comment]
@dramich We don't expect to allow users to enable monitoring charts (in fact, install any chart from Rancher) in the Harvester cluster since we will have our own monitoring solutions built-in (based on Rancher monitoring). So that part should be fine.

File: providers/k3s.go [hunk 3]
```diff
@@ -10,6 +10,14 @@ import (
 const K3s = "k3s"
 
 func IsK3s(ctx context.Context, k8sClient kubernetes.Interface) (bool, error) {
+	isHarvester, err := IsHarvester(ctx, k8sClient)
+	if err != nil {
+		return false, err
+	}
+	if isHarvester {
+		return false, nil
+	}
```
Comment (guangbochen) [inline providers/k3s.go:19]:
```
if err != nil || isHarvester{
	return false, err
}
```

File: providers/rke2.go [hunk 4]
```diff
@@ -10,6 +10,14 @@ import (
 const RKE2 = "rke2"
 
 func IsRKE2(ctx context.Context, k8sClient kubernetes.Interface) (bool, error) {
+	isHarvester, err := IsHarvester(ctx, k8sClient)
+	if err != nil {
+		return false, err
+	}
+	if isHarvester {
+		return false, nil
+	}
```
Comment (guangbochen) [inline providers/rke2.go:19]: ditto

File: providers/k3s.go (hunk 3 above)
Comment (futuretea) [reply providers/k3s.go]: fixed

**Review by guangbochen — APPROVED** [review]

**Review by gitlawr — APPROVED** [review]

File: detector.go [hunk 5]
```diff
@@ -4,8 +4,9 @@ import (
 	"context"
 	"errors"
 
-	"github.com/rancher/kubernetes-provider-detector/providers"
 	"k8s.io/client-go/kubernetes"
+
```
Comment (dramich) [inline detector.go:8]: Revert this change

**Review by dramich — COMMENTED** [review]
Small formatting change otherwise it looks good

File: detector.go (hunk 5 above)
Comment (futuretea) [reply detector.go]: fixed

**Review by dramich — APPROVED** [review]

**Comment by guangbochen** [comment]
@dramich can we tag a new release for rancher dependency bump and testing? thanks in advance.
